### PR TITLE
Provider gaposis

### PR DESCRIPTION
**Pull request by lkwdwrd**
- Group Attributes
- Default Groups
- OIDC Group Mapper support
- Identity Resource (for service users...)
- Identity Group Membership

Requires updated e3db-client-go and e3db-go versions when they are tagged and available.

### CODE REVIEW

File: docs/resources/tozny_identity.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+terraform {
+  required_version = ">= 0.13"
+  required_providers {
+    tozny = {
+      # Pull signed provider binaries from
+      # the Terraform hosted registry namespace
+      # for Tozny registry.terraform.io/tozny
+      source  = "tozny/tozny"
+      # Pin Tozny provider version
+      version = ">=0.10.1"
```
Comment (galxy25) [inline docs/resources/tozny_identity.md:22]: I think I generally leave the terraform block out of the docs, less strings to update when doing version updates at least ;-)

File: docs/resources/tozny_identity.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  ]
+  client_credentials_config = tozny_account.autogenerated_tozny_account.config
+  realm_broker_identity_credentials = tozny_realm_broker_identity.broker_identity.credentials
+  use_tozny_hosted_broker = true
+}
+
+# A resource for provisioning a TozID identity to embed into another resource.
+# Note this user specifies a password -- protect this sensitivie information
+# by ensuring state files are encrypted
+resource "tozny_identity" "machine_user_step" {
```
Comment (galxy25) [inline docs/resources/tozny_identity.md:103]: `tozny_realm_identity` for consistency with other resources please

File: docs/resources/tozny_identity_group_membership.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
+# tozny_identity_group_membership Resource
```
Comment (galxy25) [inline docs/resources/tozny_identity_group_membership.md:1]: not blocking but prefer namespacing all realm resources with `realm`, e.g. tozny_realm_identity_group_membership

File: tozny/resource_identity.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+				Description:   "The Tozny account client configuration as a JSON string",
+				Type:          schema.TypeString,
+				Optional:      true,
+				Default:       "",
+				ForceNew:      true,
+				Sensitive:     true,
+				ConflictsWith: []string{"client_credentials_filepath"},
+			},
+			"realm_name": {
+				Description: "The name of the Realm to provision the group for.",
```
Comment (galxy25) [inline tozny/resource_identity.go:39]: `identity`

File: tozny/resource_identity.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+				Description: "The base link for password resets",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"password": {
+				Description: "The password for this identity. Ideally this comes from a secret store of some kind.",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
```
Comment (galxy25) [inline tozny/resource_identity.go:72]: [`Sensitive: true`] (https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/schema#Schema.Sensitive)

File: tozny/resource_identity.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+func resourceIdentityDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	toznySDK, err := MakeToznySDK(d, m)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	realmName := strings.ToLower(d.Get("realm_name").(string))
```
Comment (galxy25) [inline tozny/resource_identity.go:144]: can we move this lowercasing to e3db-go?

File: tozny/resource_identity_group_membership.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+				Description:   "The Tozny account client configuration as a JSON string",
+				Type:          schema.TypeString,
+				Optional:      true,
+				Default:       "",
+				ForceNew:      true,
+				Sensitive:     true,
+				ConflictsWith: []string{"client_credentials_filepath"},
+			},
+			"realm_name": {
+				Description: "The name of the Realm to provision the group for.",
```
Comment (galxy25) [inline tozny/resource_identity_group_membership.go:39]: `group membership for`

File: tozny/resource_realm_default_groups.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId("")
+
+	return diags
+}
+
+func schemaToStringSlice(schemaList []interface{}) []string {
```
Comment (galxy25) [inline tozny/resource_realm_default_groups.go:145]: default to making this exported and probably relocate to provider.go?

**Review by galxy25 — APPROVED** [review]

**Review by efabens — APPROVED** [review]

File: tozny/resource_identity_group_membership.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	storedGroupList, _ := d.GetChange("group_ids")
+	groupList := schemaToStringSlice(storedGroupList.([]interface{}))
+	serverGroups, err := toznySDK.GroupMembership(ctx, identityClient.RealmIdentityRequest{
+		RealmName:  realmName,
+		IdentityID: identityID,
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	// Use a map-set to read through groups in state-order using false to indicated
```
Comment (lkwdwrd) [inline tozny/resource_identity_group_membership.go:106]: indicate

File: tozny/resource_identity_group_membership.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+	groupMapSet := map[string]bool{}
+	for _, groupID := range groupList {
+		groupMapSet[groupID] = false
+	}
+	// Update the map set with settings read from the server
+	for _, group := range serverGroups.Groups {
+		groupMapSet[group.ID] = true
+	}
+
+	// translate group list back into terraform state format
```
Comment (lkwdwrd) [inline tozny/resource_identity_group_membership.go:117]: t/T

File: tozny/resource_identity.go (hunk 6 above)
Comment (lkwdwrd) [reply tozny/resource_identity.go:144]: Will bring this up at standup. This is an e3db-clients-go method, and uses the realm domain. In doing this, I've become a bit confused as to what is the chicken and the egg with domain/name in realms... and we may be sometimes inefficiently requesting realm info at times. We should get aligned before I add logic like this to a more shared library.

**Pull request merged**